src/main/java/com/commercetools/sync/cartdiscounts/CartDiscountSync.java

```diff
@@ -0,0 +1,327 @@
+import io.sphere.sdk.cartdiscounts.CartDiscount;
+import io.sphere.sdk.cartdiscounts.CartDiscountDraft;
+import io.sphere.sdk.commands.UpdateAction;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import static com.commercetools.sync.cartdiscounts.utils.CartDiscountSyncUtils.buildActions;
+import static com.commercetools.sync.commons.utils.SyncUtils.batchElements;
+import static java.lang.String.format;
+import static java.util.Optional.ofNullable;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.Collectors.toSet;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+/**
+ * This class syncs cart discount drafts with the corresponding cart discounts in the CTP project.
+ */
+public class CartDiscountSync extends BaseSync<CartDiscountDraft, CartDiscountSyncStatistics, CartDiscountSyncOptions> {
+
+    private static final String CTP_CART_DISCOUNT_FETCH_FAILED =
+        "Failed to fetch existing cart discounts with keys: '%s'.";
+    private static final String CTP_CART_DISCOUNT_UPDATE_FAILED =
+        "Failed to update cart discount with key: '%s'. Reason: %s";
+    private static final String CART_DISCOUNT_DRAFT_HAS_NO_KEY =
+        "Failed to process cart discount draft without key.";
+    private static final String CART_DISCOUNT_DRAFT_IS_NULL =
+        "Failed to process null cart discount draft.";
+
+
+    private final CartDiscountService cartDiscountService;
+
+    public CartDiscountSync(@Nonnull final CartDiscountSyncOptions cartDiscountSyncOptions) {
+        this(cartDiscountSyncOptions, new CartDiscountServiceImpl(cartDiscountSyncOptions));
+    }
+
+    /**
+     * Takes a {@link CartDiscountSyncOptions} and a {@link CartDiscountService} instances to instantiate
+     * a new {@link CartDiscountSync} instance that could be used to sync cart discount drafts in the CTP project
+     * specified in the injected {@link CartDiscountSyncOptions} instance.
+     *
+     * <p>NOTE: This constructor is mainly to be used for tests where the services can be mocked and passed to.
+     *
+     * @param cartDiscountSyncOptions the container of all the options of the sync process including the CTP project
+     *                                client and/or configuration and other sync-specific options.
+     * @param cartDiscountService     the cart discount service which is responsible for fetching/caching
+     *                                the CartDiscounts from the CTP project.
+     */
+    CartDiscountSync(@Nonnull final CartDiscountSyncOptions cartDiscountSyncOptions,
```

Comment:
it is not but it's better to add this because it's consistent for other Sync classes, and it explains the case as a note like below:
```
NOTE: This constructor is mainly to be used for tests where the services can be mocked and passed to.
``` 

Besides that I added the Javadoc for public constructor here: efdd507

Reply:
Personally I find it quite useless to add all the obvious comments to constructor. I would leave only the note.

Reply:
the thing is we have checkstyle rule that forces javadocs for these constructors. But I think it's a good sign for the user when everything has a javadoc. Eventhough it won't add move value than the common sense. But we can't assume what is common sense for all users. We have to think that no javadoc at all might leave some users with some questions too.

In this specific use case, since the constructor is only used for tests, it's be enough to leave the note